tools/diagnostics/agent.py

```diff
@@ -0,0 +1,150 @@
+import logging
+import os
+import re
+import retrying
+from typing import List
+
+import sdk_cmd
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RETRY_WAIT = 1000
+DEFAULT_RETRY_MAX_ATTEMPTS = 5
+
+
+def is_http_server_error(http_status_code: int) -> bool:
+    return http_status_code >= 500 and http_status_code <= 599
+
+
+@retrying.retry(wait_fixed=DEFAULT_RETRY_WAIT, stop_max_attempt_number=DEFAULT_RETRY_MAX_ATTEMPTS)
+def debug_agent_files(agent_id: str) -> List[str]:
+    response = sdk_cmd.cluster_request(
+        "GET",
+        "/slave/{}/files/debug".format(agent_id),
+        retry=False,
+        raise_on_error=False,
+        log_response=False,
+    )
+
+    if is_http_server_error(response.status_code):
+        # Retry.
+        raise Exception(response)
+
+    if not response.ok:
+        # Retry.
+        raise Exception(response)
+
+    return response.json()
+
+
+@retrying.retry(wait_fixed=DEFAULT_RETRY_WAIT, stop_max_attempt_number=DEFAULT_RETRY_MAX_ATTEMPTS)
+def browse_agent_path(agent_id: str, agent_path: str) -> List[dict]:
+    response = sdk_cmd.cluster_request(
+        "GET",
+        "/slave/{}/files/browse?path={}".format(agent_id, agent_path),
+        retry=False,
+        raise_on_error=False,
+        log_response=False,
+    )
+
+    if is_http_server_error(response.status_code):
+        # Retry.
+        raise Exception(response)
+
+    if response.status_code == 404:
+        return []
+
+    if not response.ok:
+        # Retry.
+        raise Exception(response)
+
+    return response.json()
+
+
+def browse_executor_sandbox(agent_id: str, executor_sandbox_path: str) -> List[dict]:
+    return browse_agent_path(agent_id, executor_sandbox_path)
+
+
+def browse_executor_tasks(agent_id: str, executor_sandbox_path: str) -> List[dict]:
+    return browse_executor_sandbox(agent_id, os.path.join(executor_sandbox_path, "tasks"))
+
+
+@retrying.retry(wait_fixed=DEFAULT_RETRY_WAIT, stop_max_attempt_number=DEFAULT_RETRY_MAX_ATTEMPTS)
```

Comment:
`browse_task_sandbox` -> `browse_executor_tasks` -> `browse_executor_sandbox` -> `browse_agent_path`
Both the first and last function in this chain have `retrying.retry`.

Let me re-iterate my comment from July: https://github.com/mesosphere/dcos-commons/pull/2511#pullrequestreview-141811827

Reply:
Good catch, thanks! 453c929a4ece907683fcfd672132c1e5aceac324